discovery/consul/consul_test.go

```diff
@@ -46,3 +51,35 @@ func TestNonConfiguredService(t *testing.T) {
 		t.Errorf("Expected service %s to be watched", "nonConfiguredServiceName")
 	}
 }
+
+func TestConsulDiscovery(t *testing.T) {
+	srv1, err := testutil.NewTestServer()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer srv1.Stop()
+
+	srv1.AddService(t, "testService", structs.HealthPassing, []string{"master"})
+	srv1.AddCheck(t, "service:testService", "testService", structs.HealthPassing)
+
+	conf := &config.ConsulSDConfig{
+		Server:       srv1.HTTPAddr,
+		Services:     []string{"testService"},
+		WatchTimeout: 10 * time.Millisecond,
+	}
+	consulDiscovery, err := NewDiscovery(conf)
+	ctx, cancelFunc := context.WithTimeout(context.Background(), time.Millisecond*10)
```

Comment:
Yes, if we don't get a hold of the services from consul within 10 milliseconds this would result in the error you mentioned, I agree, this limit is a bit low. I'll set it to one second.
We need this though because of the implementation in consul.go. It has an exit condition in the endless loop that checks the context. This was the easiest way I found to exit that function within a guaranteed time. So basically it keeps going in the background forever otherwise.

The WatchTimeout parameter is used for blocking queries to consul. https://www.consul.io/api/index.html

"When this is provided, the HTTP request will "hang" until a change in the system occurs, or the maximum timeout is reached. A critical note is that the return of a blocking request is no guarantee of a change. It is possible that the timeout was reached or that there was an idempotent write that does not affect the result of the query."

So yeah, currently we wait for WatchTimeout seconds for every request.